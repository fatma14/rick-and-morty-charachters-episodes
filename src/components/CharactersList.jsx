import React from "react";

import Character from "./Character";

export default function CharactersList({ characters, selectCharacter }) {
  return (
    <div>
      {characters.map((character) => (
        <Character
          characterName={character.name}
          handleClick={() => {
            selectCharacter(character.id);
          }}
        />
      ))}
    </div>
  );
}
