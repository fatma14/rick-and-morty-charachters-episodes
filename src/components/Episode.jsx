import React, { Component } from "react";

import getEpisodeById from "../service/episodes";
import "./episode.css";

export default class Episode extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: "",
      code: "",
      air_date: "",
    };
  }
  componentDidMount() {
    getEpisodeById(this.props.episodeURL).then((response) => {
      this.setState({
        name: response.name,
        air_date: response.air_date,
        code: response.episode,
      });
    });
  }

  render() {
    return (
      <div className="episodeContainer">
        <h3> {this.state.name} </h3>
        <div>
          <b>Episode Code:</b> {this.state.code}{" "}
        </div>
        <div>
          <b>Air Date:</b> {this.state.air_date}{" "}
        </div>
      </div>
    );
  }
}
