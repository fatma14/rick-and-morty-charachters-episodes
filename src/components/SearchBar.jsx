import React from "react";

export default function SearchBar({ searchPhrase, handleChange }) {
  return (
    <div>
      <label>
        <input
          type="text"
          value={searchPhrase}
          onChange={(event) => {
            event.preventDefault();
            handleChange(event.target.value);
          }}
        />
      </label>
    </div>
  );
}
