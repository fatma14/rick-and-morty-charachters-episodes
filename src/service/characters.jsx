import axios from "axios";

const url = "https://rickandmortyapi.com/api/character/";

export async function getCharacters() {
  const response = await axios.get(url);
  return response.data;
}

export async function getCharacterById(id) {
  const response = await axios.get(url + `${id}`);
  return response.data;
}

export async function getCharactersByName(name) {
  const response = await axios.get(url + `?name=${name}`);
  return response.data;
}

export async function getPage(url) {
  const response = await axios.get(url);
  return response.data;
}
