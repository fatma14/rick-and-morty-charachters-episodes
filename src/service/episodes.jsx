import axios from "axios";

export default async function getEpisodeById(url) {
  const response = axios.get(url);
  return response.data;
}
