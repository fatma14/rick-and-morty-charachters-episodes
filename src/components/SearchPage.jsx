import React, { Component } from "react";

import CharactersList from "./CharactersList";
import SearchBar from "./SearchBar";
import {
  getCharacters,
  getCharactersByName,
  getPage,
} from "../service/characters";
import "./searchPage.css";

export default class SearchPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      characters: [],
      searchPhrase: "",
      prev: undefined,
      next: undefined,
    };
  }

  async componentDidMount() {
    const response = await getCharacters();
    this.updateCharacters(response);
  }

  async handleChange(value) {
    const searchPhrase = this.state.searchPhrase;
    this.setState({
      searchPhrase: value,
    });

    const response = await getCharactersByName(searchPhrase);
    this.updateCharacters(response);
  }

  updateCharacters(response) {
    this.setState({
      characters: response.results.map(({ id, name }) => ({
        id,
        name,
      })),
      prev: response.info.prev === "" ? undefined : response.info.prev,
      next: response.info.next === "" ? undefined : response.info.next,
    });
  }

  changeCharacterId(id) {
    this.props.changeCharacterId(id);
  }

  async handleClick(url) {
    const response = await getPage(url);
    this.updateCharacters(response);
  }

  render() {
    return (
      <div className="searchBox">
        <SearchBar handleChange={(value) => this.handleChange(value)} />
        <CharactersList
          characters={this.state.characters}
          selectCharacter={(...arg) => this.changeCharacterId(...arg)}
        />
        <div className="buttons">
          {this.state.prev !== undefined ? (
            <button
              className="navigationButtons"
              onClick={(event) => {
                event.preventDefault();
                this.handleClick(this.state.prev);
              }}
            >
              Prev
            </button>
          ) : null}

          {this.state.next !== undefined ? (
            <button
              className="navigationButtons"
              onClick={(event) => {
                event.preventDefault();
                this.handleClick(this.state.next);
              }}
            >
              next
            </button>
          ) : null}
        </div>
      </div>
    );
  }
}
