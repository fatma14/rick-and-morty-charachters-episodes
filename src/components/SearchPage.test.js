import React from "react";
import renderer from "react-test-renderer";
import SearchPage from "./SearchPage";
import "./Episode";
import { getPage } from "../service/characters";
import "./CharactersList";
import "./SearchBar";

let mockGetCharacters;
let mockGetCharactersByName;
let mockGetPage;

jest.mock("../service/characters", () => ({
  __esModule: true,
  getCharacters: () => mockGetCharacters,
  getCharactersByName: () => mockGetCharactersByName,
  getPage: jest.fn(() => mockGetPage),
}));

jest.mock("./CharactersList", () => ({
  __esModule: true,
  default: (props) => <mock-CharactersList {...props} />,
}));

jest.mock("./SearchBar", (props) => ({
  __esModule: true,
  default: (props) => <mock-SearchBar {...props} />,
}));

describe("SearchPage unit test", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should match snapshot without prev/next buttons", async () => {
    const characters = {
      results: [{ id: 1, name: "test" }],
      info: {
        prev: "",
        next: "",
      },
    };
    mockGetCharacters = new Promise((resolve) => {
      resolve(characters);
    });

    let component;
    await renderer.act(async () => {
      component = renderer.create(<SearchPage changeCharacterId={() => {}} />);
    });

    expect(component.toJSON()).toMatchSnapshot();
  });

  describe("Pagination", () => {
    it("should match snapshot with prev/next buttons", async () => {
      const characters = {
        results: [{ id: 1, name: "test" }],
        info: {
          prev: "prevPage",
          next: "nextPage",
        },
      };
      mockGetCharacters = new Promise((resolve) => {
        resolve(characters);
      });

      let component;
      await renderer.act(async () => {
        component = renderer.create(
          <SearchPage changeCharacterId={() => {}} />
        );
      });

      expect(component.toJSON()).toMatchSnapshot();
    });

    it("should call getPage when clicking on next Page", async () => {
      const characters = {
        results: [{ id: 1, name: "test" }],
        info: {
          prev: "prevPage",
          next: "nextPage",
        },
      };
      mockGetCharacters = Promise.resolve(characters);
      mockGetPage = Promise.resolve(characters);

      let component;
      await renderer.act(async () => {
        component = renderer.create(
          <SearchPage changeCharacterId={() => {}} />
        );
      });
      component.root
        .findAllByType("button")[0]
        .props.onClick({ preventDefault: () => {} });

      expect(getPage).toBeCalled();
    });
  });
});
