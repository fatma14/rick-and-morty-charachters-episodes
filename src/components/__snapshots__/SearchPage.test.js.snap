// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage unit test Pagination should match snapshot with prev/next buttons 1`] = `
<div
  className="searchBox"
>
  <mock-SearchBar
    handleChange={[Function]}
  />
  <mock-CharactersList
    characters={
      Array [
        Object {
          "id": 1,
          "name": "test",
        },
      ]
    }
    selectCharacter={[Function]}
  />
  <div
    className="buttons"
  >
    <button
      className="navigationButtons"
      onClick={[Function]}
    >
      Prev
    </button>
    <button
      className="navigationButtons"
      onClick={[Function]}
    >
      next
    </button>
  </div>
</div>
`;

exports[`SearchPage unit test should match snapshot without prev/next buttons 1`] = `
<div
  className="searchBox"
>
  <mock-SearchBar
    handleChange={[Function]}
  />
  <mock-CharactersList
    characters={
      Array [
        Object {
          "id": 1,
          "name": "test",
        },
      ]
    }
    selectCharacter={[Function]}
  />
  <div
    className="buttons"
  />
</div>
`;
